cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysisStage.qll

```diff
@@ -175,9 +143,137 @@ signature module UtilSig<DeltaSig DeltaParam> {
+   */
+  SemType getAlternateTypeForSsaVariable(SemSsaVariable var);
 }
 
-module RangeStage<DeltaSig D, UtilSig<D> LangParam> {
+signature module UtilSig<DeltaSig DeltaParam> {
+  SemExpr semSsaRead(SemSsaVariable v, DeltaParam::Delta delta);
+
+  SemGuard semEqFlowCond(
+    SemSsaVariable v, SemExpr e, DeltaParam::Delta delta, boolean isEq, boolean testIsTrue
+  );
+
+  predicate semSsaUpdateStep(SemSsaExplicitUpdate v, SemExpr e, DeltaParam::Delta delta);
+
+  predicate semValueFlowStep(SemExpr e2, SemExpr e1, DeltaParam::Delta delta);
+
+  /**
+   * Gets the type used to track the specified source variable's range information.
+   *
+   * Usually, this just `e.getType()`, but the language can override this to track immutable boxed
+   * primitive types as the underlying primitive type.
+   */
+  SemType getTrackedTypeForSsaVariable(SemSsaVariable var);
+
+  /**
+   * Gets the type used to track the specified expression's range information.
+   *
+   * Usually, this just `e.getSemType()`, but the language can override this to track immutable boxed
+   * primitive types as the underlying primitive type.
+   */
+  SemType getTrackedType(SemExpr e);
+}
+
+module RangeStage<DeltaSig D, LangSig<D> LangParam, UtilSig<D> UtilParam> {
+  /**
+   * An expression that does conversion, boxing, or unboxing
+   */
+  private class ConvertOrBoxExpr instanceof SemUnaryExpr {
+    ConvertOrBoxExpr() {
+      this instanceof SemConvertExpr
+      or
+      this instanceof SemBoxExpr
+      or
+      this instanceof SemUnboxExpr
+    }
+
+    string toString() { result = super.toString() }
+
+    SemExpr getOperand() { result = super.getOperand() }
+  }
+
+  /**
+   * A cast that can be ignored for the purpose of range analysis.
+   */
+  private class SafeCastExpr extends ConvertOrBoxExpr {
+    SafeCastExpr() {
+      conversionCannotOverflow(UtilParam::getTrackedType(pragma[only_bind_into](getOperand())),
+        UtilParam::getTrackedType(this))
+    }
+  }
+
+  /**
+   * A cast to a small integral type that may overflow or underflow.
+   */
+  private class NarrowingCastExpr extends ConvertOrBoxExpr {
+    NarrowingCastExpr() {
+      not this instanceof SafeCastExpr and
+      typeBound(UtilParam::getTrackedType(this), _, _)
+    }
+
+    /** Gets the lower bound of the resulting type. */
+    int getLowerBound() { typeBound(UtilParam::getTrackedType(this), result, _) }
+
+    /** Gets the upper bound of the resulting type. */
+    int getUpperBound() { typeBound(UtilParam::getTrackedType(this), _, result) }
+  }
+
+  private module SignAnalysisInstantiated = SignAnalysis<D, UtilParam>; // TODO: will this cause reevaluation if it's instantiated with the same DeltaSig and UtilParam multiple times?
+
+  private import SignAnalysisInstantiated
+
+  private module ModulusAnalysisInstantiated = ModulusAnalysis<D, UtilParam>; // TODO: will this cause reevaluation if it's instantiated with the same DeltaSig and UtilParam multiple times?
```

Comment:
A note for anyone else reading this: We discussed in our 1:1 that we believe this _should_ not cause re-evaluation because modules are applicative. But we should still make sure that we don't get re-evaluation due to different optimisation decisions.